### PR TITLE
Rewrite as_scalar_pde using syzygy modules

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

**Comment by isuruf** [comment]
cc @xywei

**Review by inducer — COMMENTED** [review]
Some review notes we wrote during our meeting.

File: sumpy/expansion/diff_op.py [hunk 1]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+    # for each column we calculate the intersection of the left modules and the
+    # right modules. This requires only $3*(n-2)$ work.
+
+    # None represents the free module of the cartesian product of $m$ number of
+    # polynomials where $m$ is the number of rows of the PDE matrix.
+    left_intersections = [None]*ncols
+    right_intersections = [None]*ncols
+
+    def intersect(a, b):
+        if a is None:
```
Comment (inducer) [inline sumpy/expansion/diff_op.py:255]:
```suggestion
        # "None" serves as the neutral element.
        if a is None:
```
(or get rid of the `None` handling?)

File: sumpy/expansion/diff_op.py [hunk 2]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+            return b
+        if b is None:
+            return a
+        return a.intersect(b)
+
+    for i in range(1, ncols):
```
Comment (inducer) [inline sumpy/expansion/diff_op.py:261]: Express as scan? (Not sure if it'd be a win.)

**Review by inducer — COMMENTED** [review]
Cooll! Good to go after these minor fixes.

File: sumpy/expansion/diff_op.py [hunk 3]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
         return res
 
 
+def convert_module_to_matrix(module, *generators):
+    import sympy
+    result = []
+    for syzygy in module:
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Done in e27280b600811fae8b2710e187325fadc08f9270

File: sumpy/expansion/diff_op.py [hunk 4]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+        row = []
+        for dmp in syzygy.data:
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Done in e27280b600811fae8b2710e187325fadc08f9270

File: sumpy/expansion/diff_op.py [hunk 5]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
         return res
 
 
+def convert_module_to_matrix(module, *generators):
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Done in 1a2cd5d8e6ba7edb7e1d474add5e860df81de814

File: sumpy/expansion/diff_op.py [hunk 6]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+    import sympy
+    result = []
+    for syzygy in module:
+        row = []
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Done in 0b909f1ed71591caa1f7f4463d85368c4e06b9b7

File: sumpy/expansion/diff_op.py [hunk 7]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+        for dmp in syzygy.data:
+            row.append(sympy.Poly(dmp.to_dict(), *generators,
+                domain=sympy.EX).as_expr())
+        result.append(row)
+    return sympy.Matrix(result)
+
+
 @memoize
+def _get_all_scalar_pdes(pde):
+    r"""
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Done in b3b6487ce31c54fcd041f605ecafaebefd072225

File: sumpy/expansion/diff_op.py [hunk 8]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+        row = []
+        for dmp in syzygy.data:
+            row.append(sympy.Poly(dmp.to_dict(), *generators,
+                domain=sympy.EX).as_expr())
+        result.append(row)
+    return sympy.Matrix(result)
+
+
 @memoize
+def _get_all_scalar_pdes(pde):
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Done in 3829bcc41f4c1f126f97dff97f1cde8a7f1e4d93

File: sumpy/expansion/diff_op.py [hunk 9]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+      \frac{\partial^2 u}{\partial x^2} + 2 \frac{\partial^2 v}{\partial x y} = 0 \\
+      3 * \frac{\partial^2 u}{\partial y^2} + \frac{\partial^2 v}{\partial x^2} = 0
+    \]
+    is converted into,
+    \[
+      \begin{bmatrix}
+        x^2   & 2xy \\
+        2y^2  & x^2
+      \end{bmatrix}
+    \]
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Done in 67c2740381491d0341bdbf8331ef53af3ac79a71

File: sumpy/expansion/diff_op.py [hunk 10]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+
+    Let $r_i$ be the rows of the above matrix.  In order find a scalar PDE for
+    the i-th component, we need to find some polynomials,
+    $a_1, a_2, \ldots, a_n$ such that the vector $\sum_i a_i r_i$ has zeros
+    except for the i-th component. In other words, we need to find a vector of
+    polynomials such that the inner product of it with each of the columns except
+    for the i-th column is zero.
+    i.e. $a_1, a_2, \ldots, a_n$ is a syzygy of all columns except for the i-th
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Done in ee6ff0d781dc1564d40e915124720f566301b1ad

File: sumpy/expansion/diff_op.py [hunk 11]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+    column.
+
+    To calculate this, we first calculate the syzygy module of each column.
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Done in 363381935a2228be32fc38a21a2e5710372b8e6b

File: sumpy/expansion/diff_op.py [hunk 12]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+    A syzygy of a column vector is a row vector which has inner product zero
+    with the column vector. A syzygy module is the module of all such syzygies
+    and is generated by a finite set of syzygies.
+    After calculating the syzygy module of each column, we intersect them together
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Done in 363381935a2228be32fc38a21a2e5710372b8e6b

File: sumpy/expansion/diff_op.py [hunk 13]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+    resulting in one module of syzgies that is finitely generated and this module
+    generates all the syzygies that when multiplied by the matrix gives a vector
+    with zeros except for the i-th component. Therefore the vector
+    $a_1, a_2, \ldots, a_n$ is in this module.
+
+    For each syzygy in the generating set, we calculate the inner product with the
+    first column to get a polynomial that represents a scalar PDE for the first
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Removed in 363381935a2228be32fc38a21a2e5710372b8e6b. This was redundant

File: sumpy/expansion/diff_op.py [hunk 14]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+    A syzygy of a column vector is a row vector which has inner product zero
+    with the column vector. A syzygy module is the module of all such syzygies
+    and is generated by a finite set of syzygies.
+    After calculating the syzygy module of each column, we intersect them together
+    resulting in one module of syzgies that is finitely generated and this module
+    generates all the syzygies that when multiplied by the matrix gives a vector
+    with zeros except for the i-th component. Therefore the vector
+    $a_1, a_2, \ldots, a_n$ is in this module.
+
+    For each syzygy in the generating set, we calculate the inner product with the
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Removed in 363381935a2228be32fc38a21a2e5710372b8e6b as it was redundant

File: sumpy/expansion/diff_op.py [hunk 15]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+    first column to get a polynomial that represents a scalar PDE for the first
+    component. When there are multiple such scalar PDEs, we want to get a
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Done in 363381935a2228be32fc38a21a2e5710372b8e6b

File: sumpy/expansion/diff_op.py [hunk 16]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
+    """
+    import sympy
+    from sympy.polys.orderings import grevlex
+    gens = [sympy.symbols(f"_x{i}") for i in range(pde.dim)]
+    gens += [sympy.symbols(f"_t{i}") for i in range(pde.total_dims - pde.dim)]
+
+    max_vec_idx = 0
+    for eq in pde.eqs:
+        for deriv_ident in eq.keys():
+            max_vec_idx = max(max_vec_idx, deriv_ident.vec_idx)
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Done in 6052412dfa84da58c0ea6cd087d6f826d5aa76f5

File: sumpy/expansion/diff_op.py (hunk 1 above)
Comment (isuruf) [reply sumpy/expansion/diff_op.py:196]: Done in 54943a3ed74f7669563c8645576810161b145968

File: sumpy/expansion/diff_op.py [hunk 17]
```diff
@@ -145,7 +145,161 @@ def to_sym(self, fnames=None):
...
 
 
+def convert_module_to_matrix(module, *generators):
+    import sympy
+    result = []
+    for syzygy in module:
+        row = []
+        for dmp in syzygy.data:
+            row.append(sympy.Poly(dmp.to_dict(), *generators,
+                domain=sympy.EX).as_expr())
```
[the remark this answers is not in the document]
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: Had to keep this as it is. SymPy Matrices with Poly elements don't work. 77c36052c463b5d273fddca3fce2d3a37f71b00c

File: sumpy/expansion/diff_op.py (hunk 2 above)
Comment (isuruf) [reply sumpy/expansion/diff_op.py]: rewrote using `itertools.accumulate`.

File: sumpy/expansion/diff_op.py [hunk 18]
```diff
@@ -145,18 +147,145 @@ def to_sym(self, fnames=None):
...
+
+    return results
+
+
+def as_scalar_pde(pde: LinearPDESystemOperator, vec_idx: int) \
+        -> LinearPDESystemOperator:
     r"""
     Returns a scalar PDE that is satisfied by the *vec_idx* component
-    of *pde*.
+    of *pde*. 123.
```
Comment (isuruf) [inline sumpy/expansion/diff_op.py:235]:
```suggestion
    of *pde*.
```

File: sumpy/expansion/diff_op.py [hunk 19]
```diff
@@ -145,18 +147,145 @@ def to_sym(self, fnames=None):
...
+    # poly is a sympy DMP (dense multi-variate polynomial)
+    # type and we convert it to a sympy expression because
+    # sympy matrices with polynomial entries are not supported.
+    return sympy.Matrix([[sympy.Poly(poly.to_dict(), *generators,
+            domain=sympy.EX).as_expr() for poly in ideal.data] for ideal in module])
+
+
 @memoize
-def as_scalar_pde(pde, vec_idx):
+def _get_all_scalar_pdes(pde: LinearPDESystemOperator) -> LinearPDESystemOperator:
```
Comment (isuruf) [inline sumpy/expansion/diff_op.py:160]:
```suggestion
def _get_all_scalar_pdes(pde: LinearPDESystemOperator) -> List[LinearPDESystemOperator]:
```

File: sumpy/expansion/diff_op.py [hunk 20]
```diff
@@ -145,18 +147,145 @@ def to_sym(self, fnames=None):
...
+        pde_dict = {
+            DerivativeIdentifier(mi, 0): sym.sympify(coeff.as_expr().simplify()) for
+            (mi, coeff) in zip(scalar_pde.monoms(), scalar_pde.coeffs())
+        }
+        results.append(LinearPDESystemOperator(pde.dim, pmap(pde_dict)))
+
+    return results
+
+
+def as_scalar_pde(pde: LinearPDESystemOperator, vec_idx: int) \
```
Comment (inducer) [inline sumpy/expansion/diff_op.py:231]:
```suggestion
def as_scalar_pde(pde: LinearPDESystemOperator, comp_idx: int) \
```

File: sumpy/expansion/diff_op.py [hunk 21]
```diff
@@ -145,18 +147,145 @@ def to_sym(self, fnames=None):
...
+    basis of the polynomials. We use the Groebner basis property that the largest
+    monomial of each polynomial generated by set of polynomials is divisible
+    by the largest monomial of some polynomial in the Groebner basis.
+    When a graded monomial ordering is used this implies that the degree of any
+    polynomial generated is greater than or equal to the degree of a polynomial
+    in the Groebner basis. We choose that polynomial as our scalar PDE.
+
     :arg pde: An instance of :class:`LinearPDESystemOperator`
     :arg vec_idx: the index of the vector-valued function that we
                   want as a scalar PDE
```
Comment (inducer) [inline sumpy/expansion/diff_op.py:287]:
```suggestion
    :arg comp_idx: the index of the component of the PDE solution vector
        for which a scalar PDE is requested.
```

File: sumpy/expansion/diff_op.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (inducer) [reply sumpy/expansion/diff_op.py]: Link the issue? (If it's quick.)

File: sumpy/expansion/diff_op.py [hunk 22]
```diff
@@ -145,18 +147,145 @@ def to_sym(self, fnames=None):
...
+
+    Let :math:`r_i` be the columns of the above matrix.  In order find a scalar PDE
+    for the i-th component, we need to find some polynomials,
+    :math:`a_1, a_2, \ldots, a_n` such that the vector :math:`\sum_i a_i r_i` has
+    zeros except for the i-th component. In other words, we need to find a vector of
+    polynomials such that the inner product of it with each of the columns except
+    for the i-th column is zero. i.e. :math:`a_1, a_2, \ldots, a_n` is a syzygy
+    of all columns except for the i-th column.
+
+    To calculate a module that annihilates all but the $i$th column, we first
```
Comment (inducer) [inline sumpy/expansion/diff_op.py:266]:
```suggestion
    To calculate a module that annihilates all but the :math:`i`th column, we first
```

File: sumpy/expansion/diff_op.py [hunk 23]
```diff
@@ -145,18 +147,145 @@ def to_sym(self, fnames=None):
...
+    calculate the syzygy module of each column. A syzygy of a column vector is a
+    row vector which has inner product zero with the column vector.
+    A syzygy module is the module of all such syzygies and is generated by a
+    finite set of syzygies. After calculating the syzygy module of each column,
+    we intersect them together except for the import $th module resulting in a
```
Comment (inducer) [inline sumpy/expansion/diff_op.py:271]: `import $th`?

File: sumpy/expansion/diff_op.py [hunk 24]
```diff
@@ -145,18 +147,145 @@ def to_sym(self, fnames=None):
...
+        2y^2  & x^2
+      \end{bmatrix}.
+
+    Let :math:`r_i` be the columns of the above matrix.  In order find a scalar PDE
+    for the i-th component, we need to find some polynomials,
+    :math:`a_1, a_2, \ldots, a_n` such that the vector :math:`\sum_i a_i r_i` has
+    zeros except for the i-th component. In other words, we need to find a vector of
+    polynomials such that the inner product of it with each of the columns except
+    for the i-th column is zero. i.e. :math:`a_1, a_2, \ldots, a_n` is a syzygy
+    of all columns except for the i-th column.
```
Comment (inducer) [inline sumpy/expansion/diff_op.py:264]:
```suggestion
    of all columns except for the :math:`i`-th column.
```

File: sumpy/expansion/diff_op.py (hunk 23 above)
Comment (isuruf) [inline sumpy/expansion/diff_op.py:271]:
```suggestion
    we intersect them together except for the :math:`i`-th module resulting in a
```

File: sumpy/expansion/diff_op.py (hunk 22 above)
Comment (isuruf) [inline sumpy/expansion/diff_op.py:266]:
```suggestion
    To calculate a module that annihilates all but the :math:`i`-th column, we first
```

File: sumpy/expansion/diff_op.py [hunk 25]
```diff
@@ -145,18 +147,145 @@ def to_sym(self, fnames=None):
...
+
+    To calculate a module that annihilates all but the $i$th column, we first
+    calculate the syzygy module of each column. A syzygy of a column vector is a
+    row vector which has inner product zero with the column vector.
+    A syzygy module is the module of all such syzygies and is generated by a
+    finite set of syzygies. After calculating the syzygy module of each column,
+    we intersect them together except for the import $th module resulting in a
+    syzygy module that when multiplied by the matrix gives a matrix
+    with zeros **EXCEPT** for the i-th column. Therefore the vector
```
Comment (isuruf) [inline sumpy/expansion/diff_op.py:273]:
```suggestion
    with zeros **EXCEPT** for the :math:`i`-th column. Therefore the vector
```

File: sumpy/expansion/diff_op.py (hunk 24 above)
Comment (isuruf) [inline sumpy/expansion/diff_op.py:264]:
```suggestion
    of all columns except for the :math:`i`-th column.
```

**Review by inducer — COMMENTED** [review]
Thanks! :tada: Just a few more, then this is good to go.

File: sumpy/expansion/diff_op.py [hunk 26]
```diff
@@ -145,18 +147,145 @@ def to_sym(self, fnames=None):
...
+        }
+        results.append(LinearPDESystemOperator(pde.dim, pmap(pde_dict)))
+
+    return results
+
+
+def as_scalar_pde(pde: LinearPDESystemOperator, vec_idx: int) \
+        -> LinearPDESystemOperator:
     r"""
     Returns a scalar PDE that is satisfied by the *vec_idx* component
```
Comment (isuruf) [inline sumpy/expansion/diff_op.py:234]:
```suggestion
    Returns a scalar PDE that is satisfied by the *comp_idx* component
```

File: sumpy/expansion/diff_op.py [hunk 27]
```diff
@@ -166,57 +295,7 @@ def as_scalar_pde(pde, vec_idx):
...
-            pde_dict = {}
-            mult = indep_row[max(indep_row.keys())]
-            for k, v in indep_row.items():
-                pde_dict[DerivativeIdentifier(mis[k], 0)] = v / mult
-            plog.done()
-            return LinearPDESystemOperator(pde.dim, pmap(pde_dict))
-
-    plog.done()
-    raise AssertionError
+    return _get_all_scalar_pdes(pde)[vec_idx]
```
Comment (isuruf) [inline sumpy/expansion/diff_op.py:298]:
```suggestion
    return _get_all_scalar_pdes(pde)[comp_idx]
```

**Comment by inducer** [comment]
Pushed a minor fix, LGTM with that. Thanks for working on this!

**Pull request merged**